File: 0053-maximum-subarray/maximum-subarray.go
func maxSubArray(nums []int) int {
	maxSub := nums[0]
	subSum := nums[0]
	for _, num := range nums[1:] {
		if subSum < 0 {
			subSum = num
		} else {
			subSum += num
		}
		maxSub = max(maxSub, subSum)
	}
	return maxSub
}

File: W1_0053-maximum-subarray/maximum-subarray.ts
function maxSubArray(nums: number[]): number {
  let maxSub = nums[0];
  let subSum = nums[0];

  for (let i = 1; i < nums.length; i++) {
    if (subSum < 0) {
      subSum = nums[i];
    } else {
      subSum += nums[i];
    }
    if (subSum > maxSub) maxSub = subSum;
  }
  return maxSub;
}
